### DISCUSSION TITLE
Can't get Flash to Einsy board to take

### DISCUSSION THREAD

**Discussion by PRGeno** [dz0ny/klipper-prusa-mk3s#32] [Q&A] [answered]
I've tried several methods to get the Klippy firmware uploaded into my Einsy Rambo board to no avail.  

I've tried doing it from the Pi command line following the standard Klipper make method.  I've also downloaded the firmware from PrusaSlicer which thinks it did it's job but the printer display gets stuck at 100% verifying. whichever method used, and or whichever firmware used, it always ends up either stuck saying it updating please wait, or in the Prusa boot loader with the 3 dots.

I'm always able to load the latest Prusa firmware via PrusaSlicer, but I can't ever get a Klippy load to boot after flashing.

Has anyone experienced this?  I'm sure I'm missing something simple, but I can't find whatever that might be.

**Comment by confuzled-dev** [comment]
Have you used the serial to USB adapter cable to flash the USB fix?

**Comment by PRGeno** [comment]
I must have missed that fix.  I'm still not sure I understand what needs to be done with this. 

 I'm able to flash the official firmware from both PrusaSlicer and the Octoprint Firmware Updater. It's that Klippy firmware that doesn't appear to finish.  That would seem to indicate flashing from the USB serial port was working ok.

Can you point me towards more information regarding this fix?

**Comment by dz0ny** [comment]
How old is your Einsy board? Which model of printer do you have? Also, be aware that until Klipper is communicating with MCU, you won't see anything useful on LCD. If I remember correctly it only shows **Starting Prusa...** or something like that.

**Comment by PRGeno** [comment]
My printer is built upon self sourced and left over parts based upon a Mk3s mounted on a Bear frame with a Bondtech Mosquito extruder.  
The Einsy Rambo was sourced in 2019.  I have 2 of these built almost identical but the second one has a Bondtech with a V6 hot end instead of a Mosquito.  Both machines have operated very well for years with the latest Bondtech version (shorter Z height) of the stock Prusa firmware, including the latest release.   

I'm trying to dig up my old USB programming adapter (I haven't needed it for anything in so long I don't remember where it is).  Any additional insight into using it to flash this "USB fix" or the firmware directly through it would be greatly appreciated.

But I'm still shocked that I can flash the stock firmware via USB/Serial via both PrusaSlicer and the Octoprint Firmware Updater plugin but not the Klippy firmware.  It seems to flash fine, but seems to get stuck after the verify is completed.

**Comment by confuzled-dev** [comment]
https://nikolak.com/klipper-on-prusa-mk3s/

second paragraph. This may be helpful

**Reply by PRGeno** [reply]
I knew I missed something.  (Now, where did I put that USB adapter?)

Thanks for all of your help

**Comment by PRGeno** [comment]
OK, I believe I went through the whole process with the USB programmer adapter, seemed to get back all the right responses to the commands and flashing and then followed the Klipper flashing procedure which seemed to do what is was supposed to do.

Unfortunately there is no happiness here.

Same annoying results of the Klippy flash getting all the way to the end, being reported as successful by the flashing device, but the Prusa screen is stuck saying it's upgrading (forever - and I've waited a long time on that screen).  

Rebooting the printer then brings up the generic Prusa screen with the three dot ellipse.

Restarting Klipper and it reports no printer is attached but the Prusa screen immediately changes to "upgrading firmware do not disconnect".

Unless anyone here can point me toward a magic step I might still be missing, I'll have to decide to either stay on the Prusa stock MK3s firmware, get another MCU to replace the Einsy, or go all in on Prusa and get a 4S (or 3.9) upgrade and make the modifications required to make the upgrade work on my Bear frame.

**Comment by dz0ny** [comment]
How are you connecting RPI to the board?

> Rebooting the printer then brings up the generic Prusa screen with the three dot ellipse.
That is the correct state of the LCD that you will see until the klipper connects to the board.

**Reply by PRGeno** [reply]
Built in USB, not the programmer adapter

**Comment by PRGeno** [comment, marked as answer]
OK, I finally got it working.  I'm reporting the problem here to hopefully help future prospective converts from making the same mistake I did.

The problem was with the the options I chose when I compiled my Klipper Firmware in the make menuconfig process.  I was too smart for my own good and set what I thought were the correct extra configuration options for the AVR atmega2560.  

I set the processor to the atmega2560 but then also set the processor speed to 8Mhz and the baud rate for the serial port to 115200 (the same as my previous Octoprint serial connection).  I left the communication interface at its default of uart0.

I suspect it was the baud rate change that was my downfall.  I went back and disabled the extra low level configuration options and saved that config again.  After verifying that it reset all those options to the default, I tried compiling and uploading again.

And what do you know?  It worked.

The moral of this story is to leave the extra low level config options at the default settings after changing the processor model to the AVR atmega2560.    

Now let's see if all this was worth the effort and if Klipper has anything to offer beyond what Prusa gave me.  The Mk3s might go back the stock Prusa firmware again if it doesn't.  This was really just an exercise to learn more about Klipper before I build my next printer, a Voron 2.4. 

Thanks again for all your help

**Reply by PapaPlintus** [reply]
Ugh I can't get this to work. My LCD screen just says it's updating the firmware and mainsail can't connect to the MCU. No idea what I'm doing wrong.

**Reply by nakwada** [reply]
@PapaPlintus I saw your post and am facing the xact same issue with my MK3S and my MK3S+.
First is a kit I built in early 2020, second is another kit from late 2021.
I suspect an issue related to the mainboard chip's firmware, as described here: https://nikolak.com/klipper-on-prusa-mk3s/

I haven't tried this procedure yet.


EDIT: lol, I'm a retard!
I went and edited the printer.cfg file as follow

```
[mcu]
# serial: /dev/serial0 # If you are using internal RPI serial port
serial: /dev/ttyACM0 # If you are using RPI via USB connection to printer
```

I commented the first line and uncommented the second. It works now!
